scripts/tutorial-uploader/lib/api.ts

```diff
@@ -0,0 +1,299 @@
+    }
+    if (ids.length === 1) {
+      return ids[0];
+    }
+    throw new Error(
+      `Found ${ids.length} items for getId('${collection}', '${field}', '${value}'. ` +
+        `Expected one or none.`,
+    );
+  }
+
+  async getTutorialIdBySlug(slug: string): Promise<string | null> {
+    return await this.getId("tutorials", "slug", slug);
+  }
+
+  /**
+   * Tutorials can have many translations, but we only use English at the moment.
+   */
+  async getEnglishTranslationId(tutorialId: string): Promise<number> {
+    const response = await this.client.request(
+      readItem("tutorials", tutorialId, {
+        // @ts-ignore
+        fields: ["translations.id", "translations.languages_code"],
+      }),
+    );
+    if (!response.translations) {
+      throw new Error(
+        `Something went wrong getting translations for tutorial ${tutorialId}`,
+      );
+    }
+    const englishTranslation = (response.translations as Translation[]).find(
+      (x: Translation) => x.languages_code === "en-US",
+    );
+    if (!englishTranslation) {
+      throw new Error(`No english translation for tutorial ${tutorialId}`);
+    }
+    return englishTranslation.id;
+  }
+
+  /**
+   * Remove all topics from a tutorial.
+   * This works by deleting appropriate entries of
+   * "tutorials_tutorials_topics": a collection of one-to-many mappings of
+   * "tutorials" to "tutorials_topics".
+   */
+  async clearTopics(tutorialId: string) {
+    // "tutorials_tutorials_topics" is mapping of tutorial to topics
+    const ids = await this.getIds(
+      "tutorials_tutorials_topics",
+      "tutorials_id",
+      tutorialId,
+    );
+    for (const id of ids) {
+      await this.client
+        .request(deleteItem("tutorials_tutorials_topics", id))
+        .catch((err: any) => console.log(err));
+    }
+  }
+
+  async updateTutorialTopics(tutorialId: string, topicNames: string[]) {
+    await this.clearTopics(tutorialId);
+    for (const name of topicNames) {
+      const id = await this.getId("tutorials_topics", "name", name);
+      if (id === null) throw new Error(`No topic with name '${name}'`);
+      await this.client.request(
+        createItem("tutorials_tutorials_topics", {
+          tutorials_id: tutorialId,
+          tutorials_topics_id: id,
+        }),
+      );
+    }
+  }
+
+  /* Returns the file's ID */
+  async uploadLocalFolder(path: string): Promise<string> {
+    // Zip folder
+    const zippedFilePath = `${tmpdir()}/${randomBytes(8).toString(
+      "hex",
+    )}/tutorial.zip`;
+    await $`mkdir -p ${dirname(zippedFilePath)}`;
+    // We cd into the parent dir to avoid including the full directory
+    // structure in the zip
+    await $`(cd ${dirname(path)} && zip -qr ${zippedFilePath} ${basename(
+      path,
+    )})`;
+
+    // Build form
+    const file = new Blob([await readFile(zippedFilePath)], {
+      type: "application/zip",
+    });
+    const formData = new FormData();
+    formData.append("title", zippedFilePath);
```

Comment:
What does this do? I'm surprised we'd want the title to have the random hex code and `tutorial.zip`

Reply:
From what I can tell, it's the human-readable title used in the CMS. Since the file gets destroyed after it's linked to the tutorial, it shouldn't matter much. If a human did see the file, I think the hex code helps convey it's for machines.